library/core/src/hint.rs

```diff
@@ -219,6 +219,73 @@ pub fn spin_loop() {
 /// backend used. Programs cannot rely on `black_box` for *correctness* in any way.
 ///
 /// [`std::convert::identity`]: crate::convert::identity
+///
+/// # When is this needed?
+///
+/// The most common use case for `black_box` is to make sure that benchmarks actually run the code
+/// they contain, rather than being allowed to optimize parts away based on context. Observe the
+/// example:
+///
+/// ```
+/// fn contains(haystack: &[&str], needle: &str) -> bool {
+///     haystack.iter().any(|x| x == &needle)
+/// }
+///
+/// pub fn benchmark() {
+///     let haystack = vec!["abc", "def", "ghi", "jkl", "mno"];
+///     let needle = "ghi";
+///     for _ in 0..10 {
+///         contains(&haystack, needle);
+///     }
+/// }
+/// ```
+///
+/// The compiler _could_ theoretically make optimizations like the following:
+///
+/// - `needle` and `haystack` are always the same, move the call to `contains` outside the loop and
+///   delete the loop
+/// - Inline `contains`
+/// - `needle` and `haystack` have values known at compile time, `contains` is always true. Remove
+///   the call and replace with `true`
+/// - Nothing is done with the result of `contains`: delete this function call entirely
+/// - Benchmark now has no purpose: delete this function
+///
+/// It is not likely that all of the above happens, but the compiler is definitely able to make
+/// some optimizations that would result in a very inaccurate benmark. This is where `black_box`
```

Comment:
I guess `benmark` is a typo?

Reply:
Whoops good catch, fixed this in my latest push